Let user opt-out of building unit tests

kcgen opened the pull request:
A feature is added to Meson allowing the user to explicitly enable or disable building of unit tests:

 - `-Dunit_tests=enabled`
 - `-Dunit_tests=disabled`

If the user doesn't specified this preference, then it defaults to `auto`-mode, which means unit tests **will be built** for all build-types except for `release`.

Configurable options list:

![2021-02-24_08-11](https://user-images.githubusercontent.com/1557255/109031296-728e9780-7679-11eb-8e84-1a12f80dec35.png)

Example configuration when testing is enabled:

![2021-02-24_08-10](https://user-images.githubusercontent.com/1557255/109031325-78847880-7679-11eb-94f9-dac0f36b7b48.png)

Example configuration when testing is disabled:

![2021-02-24_08-08](https://user-images.githubusercontent.com/1557255/109031375-86d29480-7679-11eb-9893-7f8595efa439.png)

Thank you @granminigun for this suggestion!

dreamer commented:
Before I'll start with the review…

Why do we need this? tests will be automatically disabled when `gtest` is not available and cannot be downloaded. If user does not want to install or download `gtest`, then the dependency can be skipped with `meson setup --wrap-mode=nodownload <builddir>`.

There are other relevant options as well: https://mesonbuild.com/Subprojects.html#commandline-options

kcgen commented:
> tests will be automatically disabled when gtest is not available

Because gtest is a general-purpose development library, someone might have gtest for reasons that don't involve DOSBox Staging.

A second point is my own personal use I wear two hats on my systems; one as a developer and two as a user.  

As a user, I routinely rebuild the master branch of Staging with maximum optimizations including LTO, especially on my slower systems. Building and linking unit tests adds minutes of time to these systems, especially given LTO solutions can't be ccached.  Also, LTO linking is performed in parallel based on number of cores.. but for single CPU systems this isn't an option (and we link a separate executable per unit tests). 

So this PR takes care of this by dilemma automatically: "oh, you want an optimized build and you left `-Dtests=auto`? Then we will skip the unit tests."

GranMinigun commented:
> the dependency can be skipped with `meson setup --wrap-mode=nodownload <builddir>`.

...and then I won't be able to build the project with mt32emu. Downloading all wraps, manually deleting unneded ones, and then configuring with wrap download disabled looks at least inconvenient. Looking at the provided docs, there seems to be no way of disabling fallback for gtest only. And if I have them installed system-wide, I may not necessarily want to always build all projects with tests enabled.

dreamer reviewed (changes requested):
OK, the rationale for adding this feature is good enough for me :)

@kcgen please note, that with Meson you can have multiple build directories, and you can store them outside of source code - so you can have a directory for LTO builds, directory for debug builds, separate directories per git branch (however you'll organize it is up to you). There's no need to re-run `meson setup` over and over again :)

kcgen commented:
Thanks for the feedback @dreamer; update inbound!
 - I split the two conditions into separate `if` statements; which reads easier and doesn't use temporary variables.
 - I also added a message informing the user if we when we automatically skip unit tests on release builds.

kcgen commented on tests/meson.build:
```diff
@@ -1,3 +1,16 @@
+
+if (get_option('use_gtest').auto() and
+    get_option('buildtype') == 'release')
+  message('Skipping unit tests (release-build without gtest preference)')
```
Better wording:
`Skipping unit tests, gtest auto-disabled for release builds`

dreamer commented on meson_options.txt:
```diff
@@ -65,3 +65,8 @@ option('try_static_libs',
        value : [],
        description : 'Attempt to statically link selected libraries.')
+
+option('use_gtest',
+       type : 'feature',
+       value : 'auto',
+       description : 'Build unit tests')
```
I think I made a mistake in asking for rename to `use_gtest`. All `use_*` options we have get value 'false'/'true' (and one gets 'auto'). Maybe we'll transition to use option type feature for them, but not for as long as we use 2 buildsystems.

Let's change name of this option again, to: `unit_tests`. This way user will be configuring it using `-Dunit_tests=enabled`, `-Dunit_tests=disabled`; and it will be in line with description.

dreamer commented on tests/meson.build:
```diff
@@ -1,3 +1,16 @@
+
+# gtest conditionals
+#
+if get_option('use_gtest').disabled()
+  subdir_done()
+endif
```
I didn't know about `subdir_done()` - cool.

It introduces a problem for future, still - it means all other kinds of tests must be declared above unit tests (and this might be not the best way to do organize this file). I also don't like that these two `subdir_done` make description in comment below ('gtest dependency') inaccurate.

I suggest the following change:

change code between lines 25 and 30 to:
```meson
if (get_option('buildtype') == 'release' and get_option('unit_tests').auto())
  gtest_dep = disabler()
else
  gtest_dep = dependency('gtest', main : true,
                         required : get_option('unit_tests'),
                         fallback : ['gtest', 'gtest_main_dep'])
endif
```
I think it works exactly as we need it to work.

(please note I skipped `message`s - Meson already generates good message on its own when written this way)

And then adjust the comment explaining gtest dependency to the new functionality.

kcgen replied:
Replaced as-is, and tested. Works very nice! Thanks for this improvement @dreamer .
Also updated the commit comment.

dreamer commented:
GCC5 build fail is a bit surprising, but it makes sense - kind of… I expected dependency method will return a disabled object in this case…

Anyway - for us right now it doesn't matter - GCC5 build used wrap-mode=nofallback to disable the unit tests, because we use features from newer gtest and those had some issues with GCC5. We just want to have unit tests disabled for GCC5 build - simply add `-Dunit_test=disabled` to GCC5 build (or other builds that might fail due to wrap-mode=nofallback now).

kcgen commented:
This is proving more stubborn than we expected; I've tried just `-Dunit_tests=disabled` stand-alone, and with both, `--wrap-mode=nofallback -Dunit_tests=disabled`, but even that's not enough to stop it from trying:

https://github.com/dosbox-staging/dosbox-staging/runs/1990559060?check_suite_focus=true

![2021-02-26_13-04](https://user-images.githubusercontent.com/1557255/109355003-75cd8300-7833-11eb-8da5-5d6f12958d8f.png)

![2021-02-26_13-04_1](https://user-images.githubusercontent.com/1557255/109355006-76661980-7833-11eb-90cd-9253f20f0d89.png)

``` python

else
  gtest_dep = dependency('gtest', main : true,
                         required : get_option('unit_tests'),
                         fallback : ['gtest', 'gtest_main_dep'])
endif

```

If `unit_tests` are disabled or if `fallback` isn't available then `gtest_dep` should not be available and meson should not try to build the target.

Should we file a bug report?

kcgen commented:
Also tried:

``` python
else
  gtest_dep = dependency('gtest', main : true,
                         required : get_option('unit_tests').enabled(),
                         fallback : ['gtest', 'gtest_main_dep'])
endif
```

kcgen commented:
Ok, fourth try and it's working :sweat_smile:.
Not as pretty, but only one extra line of code (however I had to re-work the comment blocks).

dreamer reviewed (approved):
Yes, I suspect there's a Meson bug in how `dependency` kwarg `required : <feature object>` interoperates with `dependency` kwarg `fallback`. I will need to do some testing to see if that's the case (don't report the bug yet, please).

> Also tried:
> 
> ```python
> else
>   gtest_dep = dependency('gtest', main : true,
>                          required : get_option('unit_tests').enabled(),
>                          fallback : ['gtest', 'gtest_main_dep'])
> endif
> ```

This has no chance of working correctly, because `enabled()` returns boolean. When boolean is passed `required`, then we can be sure that `disabler()` dependency won't be returned.

Ehh, I don't like the new solution (now the builds will fail when fallbacks are disabled in debug builds - making it problematic to packagrs), but at this point I think it will be faster if I'll make the correction.

kcgen commented:
> This has no chance of working correctly, because enabled() returns boolean. When boolean is passed to required, then we can be sure that disabler() dependency won't be returned.

This was intended. If the user opts-out or leaves it at auto then:

`get_option('unit_tests').enabled()` evaluates to `false`, so `gtest .. ` won't be a hard-requirement.

But if the user opts-in then:

`get_option('unit_tests').enabled()` evaluates to `true`, flipping `gtest` to be a hard-requirement.
If the system or internet download can't provide getst, then we want it to fail for the user.

> Ehh, I don't like the new solution (now the builds will fail when fallbacks are disabled in debug builds - making it problematic to packagrs), but at this point I think it will be faster if I'll make the correction.

Sounds good; hope you can find a nicer compromise @dreamer!
